river/Seat.zig

```diff
@@ -342,31 +351,45 @@ pub fn handleViewUnmap(self: *Self, view: *View) void {
     if (self.focused == .view and self.focused.view == view) self.focus(null);
 }
 
-/// Handle any user-defined mapping for passed keycode, modifiers and keyboard state
-/// Returns true if the key was handled
-pub fn handleMapping(
+/// Get all user-defined mappings for passed keycode, modifiers and keyboard state
+pub fn getMappings(
     self: *Self,
     keycode: xkb.Keycode,
     modifiers: wlr.Keyboard.ModifierMask,
     released: bool,
     xkb_state: *xkb.State,
-) bool {
+) !*std.SinglyLinkedList(*Mapping) {
+    const MappingList = std.SinglyLinkedList(*Mapping);
+    const list = try util.gpa.create(MappingList);
+    list.* = .{};
+    errdefer {
+        while (list.popFirst()) |node| util.gpa.destroy(node);
+        util.gpa.destroy(list);
+    }
```

Comment:
I don't think these allocations are necessary. Rather they are an indication that we should restructure the `getMappings()` API.

It seems to me that given the current usage in `Keyboard.handleKey()` we could have one function to check if any mapping is matched. If that function returns true, add the keycode to the eaten keycodes set. Then finally run something similar to the old version of this function to run all matching mappings, asserting that at least one mapping was run.